src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3139,10 +3139,24 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           val initElems = scope.elems
           // SI-5877 The decls of a package include decls of the package object. But we don't want to add
           //         the corresponding synthetics to the package class, only to the package object class.
-          def shouldAdd(sym: Symbol) =
-            inBlock || !context.isInPackageObject(sym, context.owner)
+          // SI-6734 Locality test below is meaningless if we're not even in the correct tree.
+          def shouldAdd(sym: Symbol): Boolean = {
+            def shouldAddAsModule: Boolean =
+              sym.moduleClass.attachments.get[ClassForCaseCompanionAttachment] match {
+                case Some(att) =>
+                  val cdef = att.caseClass
+                  stats.exists {
```

Comment:
Currently, the moduleclass has attachment pointing to a classdef; but meanwhile the in-flight classdef is something transformed, so `eq` fails. The test is to say, add synthetic companion only if the associated classdef is in these statements, too. (The big loop also tries to match associated synths in an ad-hoc way.)

I'll try a bigger clean-up, where the classdef bears an attachment (to mean it has an associated synthetic; similarly for other defs). Then the big loop should just traverse looking for things with synths to add; and then at the end, add whatever remaining synths belong here.

Maybe the attachment has a `Tree => Boolean` to match something invariant that says, yes, this tree is associated with me.